### PR TITLE
Allow FPU flag override

### PR DESCRIPTION
**Pull request by franksinankaya**
Provide two environment variables as CLR_FPU_FLAGOVERRIDE to specify the VFP option in use such as vfpv3-d16 and CLR_FPU_CAPABILITY bitmask to tell which precision registers are actually implemented.

value of CLR_FPU_CAPABILITY is bit 0 for half, bit 1 for single and bit 2 for double precision.

### CODE REVIEW

**Comment by franksinankaya** [comment]
@jkotas @janvorli @BruceForstall

**Comment by franksinankaya** [comment]
Use

export CLR_FPU_CAPABILITY=0x3
export CLR_FPU_FLAGOVERRIDE="vfpv3-d16"

to build with vfpv3-d16 as an example

**Comment by janvorli** [comment]
I don't like controlling the build scripts via environment variables. If we don't want to add an explicit option to our build scripts, I'd prefer using cmake properties instead. But it seems ok to add a build.sh option for this. @jkotas I am not sure what you really meant by your comment https://github.com/dotnet/coreclr/pull/27890#issuecomment-560465520 that seems to make @franksinankaya opt for the env variable way.

**Comment by franksinankaya** [comment]
I think the concern is we have too many ARMXYZ variants. I could bring the vfp options up all the way to build.sh instead of environment variables.

**Comment by jkotas** [comment]
I agree that the environment variables are not the best way to do this.

We have a generic way to pass defines from the build command line to build. It is used to control number of features contributed by community that are not turned on in the default build. For example, https://github.com/dotnet/coreclr/pull/13567

This should use the same scheme.

**Comment by franksinankaya** [comment]
These feature flags don't seem to work in compiler detection stage in "configurecompiler.cmake"

**Comment by franksinankaya** [comment]
should I reorder include(clrdefinitions.cmake) ? or create a clrcompilerdefinitions.cmake?

**Comment by janvorli** [comment]
I am not sure I understand the problem. It think that you should be able to pass an extra cmake variabled using the -cmakeargs option (e.g. -cmakeargs -DCLR_FPU_CAPABILITY=0x7 -cmakeargs -DCLR_FPU_FLAGOVERRIDE=vfpv3) and then use these cmake variables instead of the env vars in your current implementation.

**Comment by franksinankaya** [comment]
Cool, I wasn't aware of cmakeargs feature.

File: eng/common/cross/toolchain.cmake [hunk 1]
```diff
@@ -71,8 +71,19 @@ endif()
 
 if(TARGET_ARCH_NAME MATCHES "^(arm|armel)$")
   add_compile_options(-mthumb)
-  add_compile_options(-mfpu=vfpv3)
-  if(TARGET_ARCH_NAME STREQUAL "armel")
+  set(FPUFLAG vfpv3)
+  if (FEATURE_ARM_FPU_SPEC)
```
Comment (jkotas) [inline eng/common/cross/toolchain.cmake]: I think you make this ever simpler and allow FPUFLAG to be specified on the command line directly. Then the whole change needed here is just just set check for empty `FPUFLAG` around `set(FPUFLAG vfpv3)`

**Comment by franksinankaya** [comment]
new build command for vfpv3-d16:

./build.sh -cross -arm -cmakeargs -DFEATURE_ARM_FPU_CAPABILITY=0x3 -cmakeargs -DFEATURE_ARM_FPU_TYPE=vfpv3-d16 -cmakeargs -DFEATURE_ARM_FPU_SPEC=1

File: eng/common/cross/toolchain.cmake (hunk 1 above)
Comment (franksinankaya) [reply eng/common/cross/toolchain.cmake]: I can do that. any preference on names?

**Comment by jkotas** [comment]
Nit: The command line arguments for this do not have to have `FEATURE_` prefix. It is odd to call these features.

**Comment by franksinankaya** [comment]
new build command:

./build.sh -cross -arm -cmakeargs -DARM_FPU_CAPABILITY=0x3 -cmakeargs -DARM_FPU_TYPE=vfpv3-d16

**Comment by jkotas** [comment]
Looks good to me. @janvorli ?

**Review by jkotas — APPROVED** [review]

File: eng/common/cross/toolchain.cmake [hunk 2]
```diff
@@ -71,8 +71,17 @@ endif()
...
-  if(TARGET_ARCH_NAME STREQUAL "armel")
+  if (NOT DEFINED ARM_FPU_TYPE)
+    set(ARM_FPU_TYPE vfpv3)
+  endif(ARM_FPU_TYPE)
+
+  add_compile_options(-mfpu=${ARM_FPU_TYPE})
+  if (NOT DEFINED ARM_FPU_CAPABILITY)
+    set(ARM_FPU_CAPABILITY 0x7)
+  endif(ARM_FPU_CAPABILITY)
+  add_compile_options(-DCLR_FPU_CAPABILITY=${ARM_FPU_CAPABILITY})
```
Comment (jkotas) [inline eng/common/cross/toolchain.cmake]: I am wondering whether we really need `ARM_FPU_CAPABILITY`. Can we just infer it from `ARM_FPU_TYPE`?

File: src/coreclr/configurecompiler.cmake [hunk 3]
```diff
@@ -537,10 +537,18 @@ if (CLR_CMAKE_PLATFORM_UNIX)
...
+
    # Because we don't use CMAKE_C_COMPILER/CMAKE_CXX_COMPILER to use clang
    # we have to set the triple by adding a compiler argument
    add_compile_options(-mthumb)
-   add_compile_options(-mfpu=vfpv3)
+   add_compile_options(-mfpu=${ARM_FPU_TYPE})
+   if (NOT DEFINED ARM_FPU_CAPABILITY)
+     set(ARM_FPU_CAPABILITY 0x7)
+   endif(NOT DEFINED ARM_FPU_CAPABILITY)
+   add_compile_options(-DCLR_FPU_CAPABILITY=${ARM_FPU_CAPABILITY})
```
Comment (janvorli) [inline src/coreclr/configurecompiler.cmake]: Can you please change this to "add_definitions"?

File: eng/common/cross/toolchain.cmake (hunk 2 above)
Comment (franksinankaya) [reply eng/common/cross/toolchain.cmake]: AFAIK, there are too many different combinations of VFP types from arm that we'd all have to enumerate each and every single combination here (messier IMO). ARM's original idea was to define these bitmasks with _ARM_FP and have application detect the FPU capability by looking at the half, single and double precision bits. 

This works for GCC and ARMCC. Unfortunately, clang is not compliant. I tested this the other day. We are sort of emulating ARM's idea with bitmask here.

File: eng/common/cross/toolchain.cmake [hunk 4]
```diff
@@ -71,8 +71,17 @@ endif()
 
 if(TARGET_ARCH_NAME MATCHES "^(arm|armel)$")
   add_compile_options(-mthumb)
-  add_compile_options(-mfpu=vfpv3)
-  if(TARGET_ARCH_NAME STREQUAL "armel")
+  if (NOT DEFINED ARM_FPU_TYPE)
```
Comment (janvorli) [inline eng/common/cross/toolchain.cmake]: A nit, we name other cmake variables that we pass from the command line with CLR_ prefix. I would prefer to keep this convention for these variables too.

File: eng/common/cross/toolchain.cmake (hunk 2 above)
Comment (franksinankaya) [reply eng/common/cross/toolchain.cmake]: The other problem is that vfp types used for gcc and clang are not identical. you can use vfp3-d32 on gcc. You can't do this on clang for instance. I thought it is better to push this to whoever is doing the port specify what his platform supports.

**Review by janvorli — APPROVED** [review]
LGTM modulo the comments.

File: eng/common/cross/toolchain.cmake (hunk 2 above)
Comment (jkotas) [reply eng/common/cross/toolchain.cmake]: ok

**Comment by franksinankaya** [comment]
new build command: 

./build.sh  -cross -arm -cmakeargs -DCLR_ARM_FPU_CAPABILITY=0x3 -cmakeargs -DCLR_ARM_FPU_TYPE=vfpv3-d16

**Comment by franksinankaya** [comment]
any clues on what these failures are? can we retrigger the failing tests?

**Comment by jkotas** [comment]
This is flaky test. Retriggered.

**Comment by franksinankaya** [comment]
It looks like it worked.

**Comment by BruceForstall** [comment]
@franksinankaya Would it be useful to add some comments around your code changes here to explain how to invoke the build to use these new options? E.g., giving the example of
```
./build.sh -cross -arm -cmakeargs -DCLR_ARM_FPU_CAPABILITY=0x3 -cmakeargs -DCLR_ARM_FPU_TYPE=vfpv3-d16
```
and explain the bitmask? Or, maybe better, adding a small section to the document https://github.com/dotnet/runtime/blob/master/docs/workflow/building/coreclr/cross-building.md (or perhaps https://github.com/dotnet/runtime/blob/master/docs/workflow/building/coreclr/linux-instructions.md)?

**Comment by franksinankaya** [comment]
@BruceForstall : Sure, I'll post a follow-up PR

**Pull request merged**